src/lib/tp.c

```diff
@@ -0,0 +1,152 @@
+/*
+ *   Software Updater - client side
+ *
+ *      Copyright © 2018 Intel Corporation.
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation, version 2 or later of the License.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *   Authors:
+ *         Otavio Pontes <otavio.pontes@intel.com>
+ *
+ */
+
+#define _GNU_SOURCE
+
+#include <errno.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "macros.h"
+#include "tp.h"
+
+/*
+ * To keep this thread pool implementation simple, pipes were used instead of
+ * mutex locks using the fact that read() and write() operations are thread safe.
+ * So in order to schedule a task to be executed we just need to write() a task
+ * in to the pipe and the first thread to read() that data will execute the
+ * processing.
+ */
+
+struct tp {
+	int num_threads;
+	int pipe_fds[2];
+	pthread_t threads[];
+};
+
+struct task {
+	tp_task_run_t run;
+	void *data;
+};
+
+static void *thread_run(void *data)
+{
+	int *fd;
+	ssize_t r;
+	struct task task;
+
+	fd = data;
+
+	while (1) {
+		r = read(*fd, &task, sizeof(task));
+		if (r == 0) {
+			// EOF: Close thread
+			return NULL;
+		} else if (r < 0 || r != sizeof(task)) {
+			if (errno == EINTR) {
+				// Not an error, we can continue
+				continue;
+			}
+			fprintf(stderr, "Error - Thread communication failed: %d - %s\n",
+				errno, strerror(errno));
+			return NULL;
+		}
+
+		// Run task
+		task.run(task.data);
+	}
+	return NULL;
+}
+
+struct tp *tp_start(int num_threads)
+{
+	int i;
+	struct tp *tp;
+
+	tp = calloc(1, sizeof(struct tp) + num_threads * sizeof(pthread_t *));
+	ON_NULL_ABORT(tp);
+
+	tp->num_threads = num_threads;
+	if (pipe2(tp->pipe_fds, O_DIRECT | O_CLOEXEC) < 0) {
+		goto error;
+	}
+
+	// Create threads
+	for (i = 0; i < num_threads; i++) {
+		if (pthread_create(&tp->threads[i], NULL, thread_run, &tp->pipe_fds[0]) != 0) {
+			tp->num_threads = i;
```

Comment:
Is there a reason that tp->num_threads is set earlier in the function and inside the loop? Can we remove it from the loop? Also, the value i would be from 0 to (num_threads - 1). Is that correct? So if num_threads=5, the final value set in tp->num_threads would be 4.

Reply:
The num_threads is updated in the loop if pthread_create fails, so we can release the correct number of threads in tp_complete() call in error_threads goto. At this point we have 'i' threads that were successfully created.

Reply:
Ahhh, i see nvm

Reply:
You could also just do `tp->num_threads++` under the `if` statement here to just count successful cases; both work just throwing out more options :+1: